streams/src/test/java/org/apache/kafka/streams/integration/AdjustStreamThreadCountTest.java

```diff
@@ -90,24 +91,64 @@ public void setup() {
         );
     }
 
+    private void startStreamsAndWaitForRunning(final KafkaStreams kafkaStreams) throws InterruptedException {
+        kafkaStreams.start();
+        waitForStateTransition(KafkaStreams.State.RUNNING);
+    }
+
     @After
     public void teardown() throws IOException {
+        stateTransitionHistory.clear();
         purgeLocalStreamsState(properties);
     }
 
+    private void addStreamStateChangeListener(final KafkaStreams kafkaStreams) {
+        // we store each new state in state transition so that we won't miss any state change
+        kafkaStreams.setStateListener(
+            (newState, oldState) -> stateTransitionHistory.add(newState)
+        );
+    }
+
+    private void waitForStateTransition(final KafkaStreams.State expected) throws InterruptedException {
+        waitForCondition(
+            () -> !stateTransitionHistory.isEmpty() && stateTransitionHistory.contains(expected),
+            DEFAULT_DURATION.toMillis(),
+            () -> String.format("Client did not change to the %s state in time. Observed new state transitions: %s",
+                expected, stateTransitionHistory)
+        );
+    }
```

Comment:
Couldn't we simply wait for the current state to become `RUNNING`?

```suggestion
    private void waitForRunning() throws Exception {
        waitForCondition(
            () -> kafkaStreams.state() == KafkaStreams.State.RUNNING,
            DEFAULT_DURATION.toMillis(),
            () -> String.format("Client did not transit to state %s in %d seconds", expected, DEFAULT_DURATION.toMillis() / 1000)
        );
    }
```

Reply:
We can't just check the current state to become `RUNNING` because after we add/remove threads, the state won't change immediately. That is, if we check if the state is `RUNNING` after adding/removing threads, the check will pass, but the rebalance is not happening, yet, which will cause the test fail. So I still use `stateTransitionHistory` to check the state, and I checked the last state of the history to see if it is RUNNING. That should be better.